pkg/signup/service/signup_service.go

```diff
@@ -359,6 +359,37 @@ func (s *ServiceImpl) DoGetSignup(provider ResourceProvider, userID, username st
 		return nil, nil
 	}
 
+	// Check the user_id and account_id annotations in the retrieved UserSignup.  If either of them are empty, but the
+	// values exist within the claims of the current user's Access Token then set the values in the UserSignup and update
+	// the resource.
+	userIDValue, userIDFound := userSignup.Annotations[toolchainv1alpha1.SSOUserIDAnnotationKey]
+	accountIDValue, accountIDFound := userSignup.Annotations[toolchainv1alpha1.SSOAccountIDAnnotationKey]
+
+	updateUserSignup := false
+
+	if !userIDFound || userIDValue == "" {
+		userID := ctx.GetString(context.UserIDKey)
+		if userID != "" {
+			userSignup.Annotations[toolchainv1alpha1.SSOUserIDAnnotationKey] = userID
+			updateUserSignup = true
+		}
+	}
+
+	if !accountIDFound || accountIDValue == "" {
+		accountID := ctx.GetString(context.AccountIDKey)
+		if accountID != "" {
+			userSignup.Annotations[toolchainv1alpha1.SSOAccountIDAnnotationKey] = accountID
+			updateUserSignup = true
+		}
+	}
+
+	if updateUserSignup {
+		userSignup, err = s.UpdateUserSignup(userSignup)
+		if err != nil {
+			return nil, errs.Wrapf(err, "error when attempting to update UserSignup with updated userID or accountID: %s", userSignup.Status)
```

Comment:
There is a chance that the update will return 409 here if the UserSignup was updated after we obtained it in the line 348. In this case we just need to retry instead of failing. I'm afraid that UI can start showing errors for these cases making our API less stable.

I see a few options of how we could mitigate it:
- Log 409 errors here but do not return an error. We will try to update the annotation on the next GET /signup call again.
- Use a poll loop to update the UserSignup with some short timeout. We could even do the update in a separate go routine but it could potentially cause issues in other places since we are increasing chances of getting 409 in other places...
- Combination of those two ^^^

Reply:
I've added a loop around the fetch/compare/update operation on the UserSignup, which will retry if there is a failure in updating the UserSignup and give up and return an error after repeated failed attempts.

Reply:
btw, there is already logic that does the same: https://github.com/codeready-toolchain/registration-service/blob/1c1f3bc60765ae8cf377ad9fdea1becd76740492/pkg/verification/service/verification_service.go#L163-L187